pkgs/tools/networking/openvpn3/default.nix

```diff
@@ -0,0 +1,117 @@
+{ lib
+, stdenv
+, fetchFromGitHub
+, asio
+, autoconf
+, autoconf-archive
+, autoreconfHook
+, dbus-python
+, docutils
+, glib
+, gobject-introspection
+, gtest
+, jinja2
+, jsoncpp
+, libcap_ng
+, libnl
+, libuuid
+, lz4
+, openssl
+, pkg-config
+, protobuf
+, pygobject3
+, tinyxml-2
+, wrapGAppsHook
+, wrapPython
+}:
+
+let
+  # Original repository has three submodules: openvpn3-core, asio and
+  # googletests.It's a huge pain, so we fetch openvpn3-core by hand
+  # and use the existing derivation for googletests. asio just works,
+  # apparently.
+  openvpn3-core = fetchFromGitHub {
+    owner = "OpenVPN";
+    repo = "openvpn3";
+    rev = "c4fa5a69c5d2e4ba4a86e79da8de0fc95f95edc3";
+    sha256 = "sha256-VhQkx35JKNqXKgg4i+/aJYIg3iXPGlC57wDrjDpvTyE=";
+  };
+in
+stdenv.mkDerivation rec {
+  pname = "openvpn3";
+  version = "2022-04-25";
+
+  # fetch master, because it doesn't hardcode gio-unix-2.0
+  src = fetchFromGitHub {
+    owner = "OpenVPN";
+    repo = "openvpn3-linux";
+    rev = "7bf82a0530c7865e9cb65c9d0bd0200bd1957c1d";
+    sha256 = "sha256-JIBuzuZ3yrNtcXGxJeBBx+uRJE2M+infCRIGFCPjKpw=";
+  };
+
+  postPatch = ''
+    substituteInPlace ./update-version-m4.sh --replace 'VERSION="$(git rev-parse --symbolic-full-name HEAD | cut -d/ -f3-)_$(git rev-parse --short=16 HEAD)"' "VERSION=v$version"
+    ./update-version-m4.sh
+
+    rm -r ./vendor/googletest
+    cp -r ${gtest.src} ./vendor/googletest
+    rm -r ./openvpn3-core
+    cp -r ${openvpn3-core} ./openvpn3-core
+
+    chmod -R +w ./.
+    patchShebangs .
+    echo "3.git:HEAD:c4fa5a69" > openvpn3-core-version
```

Comment:
If this is not a fake version, maybe add a comment next to `rev` so that it's updated as well when someone else will bump the version

Reply:
Fair point. I've tested it again and noticed that my derivation doesn't have correct versions, mostly. It still builds and works, but it's a fragile system that is prone to dying once the update rolls out, so I'll try to enter the version properly.

Reply:
I'm not sure to understand you here. How is this fragile ?
I prefer having the comment so that if someone else updates your derivation (which may happen pretty easily with the updates bots we have on nixpkgs), the human verifying the diff sees that it's missing something.

Reply:
About "fragile", I was referring to the fact that it is done by hand. The version setting should really be automated, which is a bit of a pain in this case.
I've worked on it for a bit and discovered another approach. We can actually build `v17_beta` instead of `master`, and then just backport a crucial commit with a patch. I think that makes more sense.
Should I submit those changes in a different commit, or should I amend the commit and force push?

Reply:
Hence the comment being a good start imo. :) (I've seen a few other packages working that way)
I'd prefer amend and force push, but ultimately I don't think it matters that much since it's merged and squashed in the end.

Reply:
I agree using 17_beta plus a patch seems better. (And if you can add a comment saying that the patch may be removed once 18_beta is released if it can be removed)